### PR TITLE
Update and delete file

### PR DESCRIPTION
**Pull request by Matt-Warnock**

### CODE REVIEW

File: lib/file_database.rb [hunk 1]
```diff
@@ -18,21 +18,40 @@ def database_empty?
...
+
+  def update(index, new_data)
+    file_array = all
+
+    file_array[index].update(new_data)
+    write_to_file(file_array)
+  end
+
+  def delete(index)
+    file_array = all.delete_at(index)
```
Comment (octopusinvitro) [inline lib/file_database.rb:40]: doesn't `delete_at` return the element that was deleted, rather than the array without the element :thinking:

File: spec/file_database_spec.rb [hunk 2]
```diff
@@ -65,6 +65,36 @@
     end
   end
 
+  describe '#contact_at' do
+    it 'takes an index and returns the contact in that index' do
+      database.create(test_details)
+      database.create(second_contact)
+
+      expect(database.contact_at(0)).to eq(test_details)
+    end
```
Comment (octopusinvitro) [inline spec/file_database_spec.rb:74]: Is a good practice (when possible) to not use other methods of the class you are testing in the test of a method, so that the tests are truly independent from each other.

For example, if something changed in the implementation of `create`, you don't want the test for `contact_at` to break, as it is unrelated.

Instead, use your method `file.write(create_json_contact)`

File: lib/file_database.rb [hunk 3]
```diff
@@ -18,21 +18,40 @@ def database_empty?
 
   def create(contact)
     file_array = all.push(contact)
-
-    file.rewind
-    file.truncate(0)
-    file << file_array.to_json
-    file.flush
+    write_to_file(file_array)
```
Comment (octopusinvitro) [inline lib/file_database.rb:21]: :+1:

File: lib/file_database.rb (hunk 1 above)
Comment (Matt-Warnock) [reply lib/file_database.rb:40]: ha, my test was that bad it didn't even see that

File: spec/file_database_spec.rb [hunk 4]
```diff
@@ -86,17 +85,16 @@
...
 
-      expect(database.all[0]).to_not eq(test_details)
+      expect(database.all.first).to eq(second_contact)
     end
   end
 
-  def create_json_contact
-    JSON.generate([test_details])
+  def create_json_contact(*contact_hash)
+    JSON.generate(contact_hash)
```
Comment (octopusinvitro) [inline spec/file_database_spec.rb:115]: Nice use of the splat operator! :+1: 
Although you can just pass an array too:

```ruby
JSON.generate([1, 2])
=> "[1,2]"

[1, 2].to_json
=> "[1,2]"
```

File: spec/file_database_spec.rb [hunk 5]
```diff
@@ -93,6 +93,24 @@
     end
   end
 
+  describe '#search' do
+    it 'returns any contacts that matches string given' do
```
Comment (octopusinvitro) [inline spec/file_database_spec.rb:97]: what happens if no contacts match the string given?
remember to always have tests for the edge cases

File: spec/file_database_spec.rb [hunk 6]
```diff
@@ -65,8 +65,54 @@
     end
   end
 
-  def create_json_contact
-    JSON.generate([test_details])
+  describe '#contact_at' do
+    it 'takes an index and returns the contact in that index' do
```
Comment (octopusinvitro) [inline spec/file_database_spec.rb:69]: What happens if there is nothing at that index? 
Will the database explode or will it handle it gracefully?

Same for the `update` and `delete` tests.

Always add tests for both the "happy path" and the "not so happy path".

File: lib/file_database.rb [hunk 7]
```diff
@@ -37,7 +37,9 @@ def update(index, new_data)
   end
 
   def delete(index)
-    file_array = all.delete_at(index)
+    file_array = all
+
+    file_array.delete_at(index)
     write_to_file(file_array)
```
[the remark this answers is not in the document]
Comment (Matt-Warnock) [reply lib/file_database.rb]: errr, that wont do anything
